fvm: remove error tracing from the kernel

Stebalien opened the pull request:
1. Remove error recording logic from the kernel entirely.
2. Remove error recording methods from the call manager.

This simplifies both and may allow us to store additional information now that the call manager handles everything internally.

TODO:

- Propagate abort messages.
- Consider recording additional information.

Stebalien commented:
@raulk I'm going to take a cleanup pass but I wanted feedback before I took this any further. Basically, it has always annoyed me how:

1. The kernel returns some error.
2. We turn back around and tell the kernel to store the error.
3. The kernel just tells the call manager to store the error.

Now, we can bypass the kernel entirely and the kernel can focus on filecoin relevant function instead of error tracing.

raulk reviewed (commented):
This is LGTM so far, @Stebalien.
- Detaching the error form the kernel makes a ton of sense, as it belongs to the outer context (call manager).
- `InvocationData` as a term is a good choice too, as it aligns well with `InvocationContainer`.
- I like encapsulating the behaviour in a Backtrace object.

In a nutshell, big thumbs up.

raulk commented:
> Consider recording additional information.

What additional info are you thinking of?

Stebalien commented:
> What additional info are you thinking of?

- Message/syscall parameters.
- Maybe a backtrace if an actor aborts (when debugging is enabled). I'm not sure what we can do here but it looks like wasmtime might capture something in the trap.

But not in _this_ PR.

Stebalien commented:
Ok, so I ended up doing a few things:

1. I did add the method number & params so we now get a nice stack trace.
2. I've switched to stack frames (actor exits) and an ultimate cause (syscall error).
3. I've introduced a new `Abort` type for passing aborts through wasmtime. This forced me to restructure the send logic a bit, but doing so caught a bug (we were returning a syscall error instead exiting with a system exit code if an actor returned an invalid block handle as the return value).

Stebalien commented on fvm/src/call_manager/default.rs:
```diff
@@ -285,57 +277,113 @@ where
+            // Store parameters, if any.
+            let param_id = if params.len() > 0 {
+                match store.data_mut().kernel.block_create(DAG_CBOR, params) {
+                    Ok(id) => id,
+                    // This could fail if we pass some global memory limit.
+                    Err(err) => return (Err(err), store.into_data().kernel.take()),
```
Having to do this manually is _kind_ of annoying, but I like having a clear cutoff where consider the syscall to have succeeded.

Stebalien commented on fvm/src/call_manager/default.rs:
```diff
@@ -285,57 +277,113 @@ where
+                    .map_err(Abort::Fatal)?;
+
+                // Lookup the invoke method.
+                let invoke: wasmtime::TypedFunc<(u32,), u32> = instance
```
I just moved the type up here and made it explicit because I get nervous about these things.

raulk replied:
Paranoia pays off ;-)

Stebalien commented:
(note: we should squash before merging)

raulk commented on fvm/src/executor/mod.rs:
```diff
@@ -43,18 +44,23 @@ impl ApplyRet {
+        bt.set_cause(Cause {
+            module: "system",
+            function: "execute",
+            error: ErrorNumber::IllegalOperation,
```
This is a bit weird, as we're picking this error from the syscall namespace?

Stebalien replied:
Yeah, I agree. I think the answer is to introduce a new type:

```rust
enum MessageFailure {
     PreValidation(...),
     Failure(Backtrace),
}
```
```

Stebalien replied:
How does it look now?

raulk commented on fvm/src/syscalls/vm.rs:
```diff
@@ -1,50 +1,42 @@
+use crate::kernel::{ClassifyResult, Context as _};
 use crate::Kernel;
 
+/// An unihabited type. We use this in `abort` to make sure there's no way to return without
```
nit

```suggestion
/// An uninhabited type. We use this in `abort` to make sure there's no way to return without
```

Stebalien commented:
And... I forgot to squash. Oh well.
